src/Components/Endpoints/test/EndpointHtmlRendererTest.cs

```diff
@@ -1178,6 +1179,59 @@ public async Task DoesNotEmitNestedRenderModeBoundaries()
         Assert.Equal("<h1>This is InteractiveWithInteractiveChild</h1>\n\n<p>Hello from InteractiveGreetingServer!</p>", prerenderedContent.Replace("\r\n", "\n"));
     }
 
+    [Fact]
+    public async Task CanSuppressCallSiteRenderModesInRootComponent()
+    {
+        // Arrange
+        var httpContext = GetHttpContext();
+        var writer = new StringWriter();
+        renderer.SuppressRootComponentRenderModes();
+
+        // Act
+        var result = await renderer.PrerenderComponentAsync(httpContext, typeof(StaticPagesRoot),
+            null,
+            ParameterView.Empty);
+        await renderer.Dispatcher.InvokeAsync(() => result.WriteTo(writer, HtmlEncoder.Default));
+        var content = writer.ToString();
+
+        // Assert
+        var numMarkers = Regex.Matches(content, MarkerPrefix).Count;
+        Assert.Equal(4, numMarkers); // A start and an end marker for each of two interactive components
+
+        var matches = Regex.Matches(content, PrerenderedComponentPattern, RegexOptions.Singleline);
+        Assert.Collection(matches,
+            match =>
+            {
+                // This match will be the grandchild
+                Assert.True(match.Success);
+                var preamble = match.Groups["preamble"].Value;
+                var preambleMarker = JsonSerializer.Deserialize<ComponentMarker>(preamble, ServerComponentSerializationSettings.JsonSerializationOptions);
+                Assert.NotNull(preambleMarker.PrerenderId);
+                Assert.Equal("server", preambleMarker.Type);
+
+                // The key thing we're observing here is that *only* the grandchild has interactivity markers.
+                // The root component and the child component are both static, even though the root component sets
+                // @rendermode=InteractiveServer on the child. This is suppressed by SuppressRootComponentRenderModes.
+                var prerenderedContent = match.Groups["content"].Value;
+                Assert.Equal("[Grandchild: Hello!]", prerenderedContent.ReplaceLineEndings(string.Empty));
+            },
+            match =>
+            {
+                // This match will be the top-level InteractiveGreetingWebAssembly
+                // The fact that interactivity is *not* suppressed here shows that SuppressRootComponentRenderModes
+                // only suppresses *call-site* rendermodes in the root. It does not stop a child from being interactive
+                // if it has a @rendermode on the child component type itself.
```

Comment:
> It feels unsound to expect that everyone is going to structure their components the way our project templates do. 

I get that and agree with it. The reason I'm OK with it here is that:

1. People don't have to use this feature. If they want to control stuff in more custom ways they can do so by setting `@rendermode=SomeExpression` like the auth templates do, and pass that through in whatever way they want.
2. It's a clear, logically coherent definition of the feature that it suppresses rendermodes in the root component, and this covers what people are asking for without requiring any app upgrade steps in typical cases.
3. We have to have *some* way to distinguish which component instances you want to force to be static and which you don't (e.g., arbitrary interactive components embedded as children into your static pages), and we want to achieve that without forcing extra configuration or upgrade steps.

> Should we now raise some sort of error if we detect a non-root-component with a @rendermode attribute that includes the Router component if there are any static routes?

TBH I don't think that's realistic. It would be much too coupled to the `Router` component and we have no way to know whether a given use of `Router` is even involved in this pattern.

> Or could we suppress any @rendermode that affects a Router parent whether or not it's in the root component when rendering a @staticpage?

I'm afraid we render top-down so we don't know what will end up in the descendant hierarchy - arbitrary code could include/exclude a Router component dynamically.

> It seems to me that whether or not the @rendermode affects the Router is more fundamental to making static routes work than whether or not the @rendermode exists in the root component.

It's not only about the router. For example in the most common scenario it's also about `HeadOutlet`.

I'm well aware this design targets a particular pattern, but they payoff from doing so is that there aren't any complicated rules. If our simple rule ("suppress rendermode in root component") works for your app then great, and if not, you have to find your own other solution. If there was a mechanism that correctly interpreted what kind of interactivity suppression patterns everyone would want in all cases, and it was trivial to explain and didn't require more code in people's apps, and was not highly expensive to implement, we'd certainly do that but keeping the feature simple and restrictive looks like the best way forwards to me.

Reply:
I tend to have a perspective similar to @halter73's on this one.

> We have to have _some_ way to distinguish which component instances you want to force to be static and which you don't (e.g., arbitrary interactive components embedded as children into your static pages), and we want to achieve that without forcing extra configuration or upgrade steps.

While the rule is simple and coherent, it's practically invisible to customers, which makes its implied limitations also invisible. Customers may perform a seemingly harmless refactor that breaks their app, and nothing in their code will suggest the reason why that happened. For the sake of comparison, "option 2" (or one of its alternatives) from your original proposal would make likely the mechanics of the feature much clearer, which IMO was one of the positive aspects of it, despite it requiring an extra upgrade step.

> People don't have to use this feature

Agreed, although I'm not sure yet what results in maximizing net customer happiness. Does the joy people get from using this feature without any configuration offset the sadness that a smaller number of customers face when they discover that they've fallen off the cliff of not being able to use the feature anymore? I'd personally rather be the customer that has to follow some extra steps to enable the feature than the one who spends an hour trying to figure out down the road why my refactor broke things.

All that said, there currently don't seem to be any [open-source examples](https://github.com/search?type=code&auto_enroll=true&q=%3CRoutes+%40rendermode+NOT+path%3AApp.razor+language%3AHTML%2BRazor&l=HTML%2BRazor) of deciding the global render mode outside the root component, so maybe this hypothetical refactor actually wouldn't be that common. So while I think it would be clearer if `@staticpage` meant "ignore all render modes" or if we took a more explicit approach like option 2, I don't actually have any evidence that the current design will cause real confusion among customers.

So I'm a currently a bit torn on what the right direction is, but I trust @SteveSandersonMS's analysis on this and am happy with proceeding with the current implementation.

Reply:
Thanks for the extra thoughts. I recognize there's some scepticism from folks about the simplifications and constraints involved in this design, and if the team isn't happy with it then we can hold back. It's not desirable to push this through if the team isn't on board with it. I'm a bit uncertain whether people are just each having to go through their own processes to internalize all the tradeoffs, or if I'm just pushing in a direction people can't buy into.

At the same time I don't see other good end-to-end design suggestions that fit the real scenarios without transferring pain to app developers. Everything within the option 2 category, to me, is like saying "*if app developers are willing to write a bunch more code and remember futher restrictions about rendermode combinations, they get more flexibility*" - which is true, but we should be hearing the message quite clearly that people are not looking to do that.

Some possible tweaks or middle-grounds that could exist:

 * We could go back to the rule that "`@staticpage` means literally everything is static, i.e., disables rendermodes completely".
   * I'm not convinced that meets the goals here, since people are asking for a way to do per-page SSR within a globally-interactive site. And SSR normally entails the option to embed interactive components if you want. Nobody has said their goal is to **block** interactivity, but rather it's to do SSR-style pages selectively.
   * However if I'm missing reasons to think that's what people are actually aiming to do, please let me know!
 * We could find a name that more literally communicates what is being done here, e.g., `@disablerootinteractivity` or `@forcestaticroot`.
   * If we can find a better name than `@staticpage` that would be great. I'm definitely not fixated on the current name - it's passable at best.
   * Names that rely on three words together is a bit much to read, given that the lowercase, no-separators naming convention. If we can find a one or two-word name, certainly better!

@MackinnonBuck @halter73 @javiercn Please let me know what product direction feels right to you! Do bear in mind the delivery capacity constraints we have to work in, too.

Reply:
> We could find a name that more literally communicates what is being done here

Yeah, I think if we could find a name that clearly conveyed what this feature was doing, that might be enough to address some of the concerns I was expressing. I'm just not sure if a name exists that's both descriptive and short enough to conveniently type and read.

I generally still can't quite push myself over the edge of being comfy with the framework ignoring the `@rendermode` gesture in such a specific manner. But I'm totally willing to be outvoted if others can get on board with it.

Reply:
One other thing that just came to mind:

There might be a scenario where a developer wants a page to be routed statically, but there's another component in the top-level layout of the app that should always be interactive (e.g., a "chat" button/window that's always there regardless of how the current page was routed). I don't have a sense of how common it would be to place a component like that in `App.razor` instead of a layout component, but that's a case where debugging could be very confusing if the developer does not understand when render modes get ignored.

Reply:
> Yeah, I think if we could find a name that clearly conveyed what this feature was doing, that might be enough to address some of the concerns I was expressing.

The most explicit but plausibly short name I can think of is `@disablerootrendermode`.

> The approach still obviously has the downside that it requires changes to application code

I personally think that any design that requires upgrade steps for existing projects, and complicates the default template for new projects, is undesirable. There's pretty strong signal from the community about that.

We're basically in a corner where it seems that we can't simultaneously satisfy all of these design goals:

 * [a] Allows individual pages to opt out of global interactivity
 * [b] Does not stop you from making other components interactive
 * [c] Does not require upgrade steps or new complications to the default template
 * [d] Is not specific to particular architectural patterns (e.g., the "global interactivity" pattern)

The design in this PR satisfies [a-c] but not [d]. Suggestions from others do satisify [d] but not [c].

I'm not sure it's possible to satisify [a-d] simultaneously in any design, since the "global interactivity" pattern in existing apps inherently specifies a rendermode for all pages at the root, and so opting out would innately mean overriding that in some sense (i.e., we have to know to disregard at least the root rendermode). Nothing in the "option 2" space solves this without forcing more code/conditions in some combination of `Program.cs` and `App.razor`, violating constraint [c].

> There might be a scenario where a developer wants a page to be routed statically, but there's another component in the top-level layout of the app that should always be interactive (e.g., a "chat" ...)

That would generally be fine since you're only suppressing `@rendermode=` in the root component. The developer is still free to put a `@rendermode` directive inside `Chat.razor` and that won't be suppressed.

I do agree that the distinction between *which* rendermodes are suppressed and which aren't wouldn't be obvious at least with the original name. Renaming it to `@disablerootrendermode` would act as a fairly strong reminder even if it still isn't 100% obvious at first glance.